Develop

Qizot opened the pull request.

Qizot committed:
Inital commit

Qizot committed:
Added partially working entry tests

Qizot committed:
Renamed 'provider' to 'adapter' and added tests

Qizot committed:
Added websockex adapter tests

Qizot committed:
Improved docs and formatted code

Hajto commented on lib/adapter/adapter.ex:
```diff
@@ -0,0 +1,83 @@
+  @type status_receiver_t :: pid()
+
+  @doc """
+  Creates new websocket connection.
```
```suggestion
  Creates a new websocket connection.
```

Hajto commented on lib/adapter/adapter.ex:
```diff
@@ -0,0 +1,83 @@
+  @doc """
+  Creates new websocket connection.
+
+  The callback should be blocking and eventually return valid connection or error on failure.
```
I would rephrase that to synchronously return new connection or failure, or something along these lines.

Hajto commented on lib/adapter/adapter.ex:
```diff
@@ -0,0 +1,83 @@
+  The callback should be blocking and eventually return valid connection or error on failure.
+
+  ## Arguments
+  - `url` - valid websocket url
```
you might as well give an example here

Hajto commented on lib/adapter/adapter.ex:
```diff
@@ -0,0 +1,83 @@
+
+  ## Arguments
+  - `url` - valid websocket url
+  - `message_receiver` - pid of process to which respond with messages incomming from socket and notify about status changes
```
I would rephrase this also. 

Grammarly suggests:
```
pid of process to which respond with messages incoming from the websocket and notify about status changes
```

I don't think that this sentence is clean and easy to understand. I would much rather prefer phrasing like `pid of incoming messages and status changes recipient`

Hajto commented on lib/adapter/adapter.ex:
```diff
@@ -0,0 +1,83 @@
+
+  """
+
+  @type connection_t :: pid()
```
I would prefer just `websocket` we have `Janus.Connection` in the ecosystem scope. I would prefer to not use an ambiguous connection if possible, and `websocket` is also as understandable.

Hajto commented on lib/adapter/adapter.ex:
```diff
@@ -0,0 +1,83 @@
+  @doc """
+  Sends payload via previously created connection.
+
+  The callback should be blocking.
```
This is not clear enough. The sending itself might be synchronous but the response will arrive asynchronously and putting `should be blocking` here might be somewhat misleading.

Hajto commented on lib/adapter/adapter.ex:
```diff
@@ -0,0 +1,83 @@
+  """
+  @spec forward_response(message_receiver_t(), payload_t()) :: :ok
+  def forward_response(message_receiver, payload) when is_pid(message_receiver) do
+    Kernel.send(message_receiver, {:ws_message, payload})
```
you can use just `send/2`, no need to specify Kernel explicitly.

Qizot replied:
`Adapter` modules contains `send/2` callback that interferes with Kernel's implementation.

Hajto commented on lib/adapter/adapter.ex:
```diff
@@ -0,0 +1,83 @@
+  @doc """
+  Helper function to forward message received via websocket to message reciever previously initialized during `c:connect/3`.
+  """
+  @spec forward_response(message_receiver_t(), payload_t()) :: :ok
```
Send returns whatever it sends so, the return is not `:ok`, but `payload_t` in this case
https://hexdocs.pm/elixir/Kernel.html#send/2

Hajto commented on lib/adapter/adapter.ex:
```diff
@@ -0,0 +1,83 @@
+  @doc """
+  Helper function to forward message received via websocket to message reciever previously initialized during `c:connect/3`.
+  """
+  @spec forward_response(message_receiver_t(), payload_t()) :: :ok
```
I am pretty sure this is library specific. So I don't think it belongs here.

Qizot replied:
`forward_response` and `notify_status` should be used by modules using `Adapter` behaviour.

Hajto commented on lib/adapter/adapter.ex:
```diff
@@ -0,0 +1,83 @@
+  @doc """
+  Helper funciton to notify given receiver with connection status change.
+  """
+  @spec notify_status(status_receiver_t(), {atom(), any}) :: :ok
```
This also looks like something that is proprietary and not reusable.

Qizot replied:
Just like above, it should be used by modules implementing `Adapter` behaviour. Maybe I should change **helper** to something more informative indicating that it should be used by adapters?

Hajto commented on lib/transport/ws.ex:
```diff
@@ -0,0 +1,69 @@
+        _timeout,
+        state(connection: connection, adapter: adapter) = s
+      ) do
+    withl encode: {:ok, payload} <- Jason.encode(payload),
```
Will the error result of `Jason.encode` ever be handled? If this fails then it is a bug in the software and it probably can't be handled in a decent way. I would consider replacing this with `Jason.encode!`

Hajto replied:
If you end up using `encode!` use `with` instead.

Hajto commented on test/adapters/websocex_test.exs:
```diff
@@ -0,0 +1,98 @@
+      end
+
+      test "return error on invalid url", %{} do
+        assert {:error, _} = Adapter.WebSockex.connect("invalid_url", self(), [])
```
I am pretty sure that if you are testing a specific library integration, with specific input, testing a specific behavior you are probably looking for a specific error not an `_`.

Hajto commented on test/adapters/websocex_test.exs:
```diff
@@ -0,0 +1,98 @@
+      end
+
+      test "return error on connection failure" do
+        assert {:error, _} = Adapter.WebSockex.connect("ws://no_server", self(), [])
```
Same as above.

Hajto commented on test/adapters/websocex_test.exs:
```diff
@@ -0,0 +1,98 @@
+      end
+    end
+
+    describe "Janus.Transport.WS when used with websocex should" do
```
```suggestion
    describe "Janus.Transport.WS when used with WebSockex should" do
```

Hajto replied:
You should also put here
```
setup do
  {:ok, {:state, connection, dunno} = state} = WS.connect({url, Adapter.WebSockex, []})
  [put values here]
end
```

Hajto commented on test/adapters/websocex_test.exs:
```diff
@@ -0,0 +1,98 @@
+if Code.ensure_loaded?(WebSockex) do
```
This clause shouldn't be necessary since this is always present in the test environment.

Hajto replied:
This file is named websocex which should be websockex

Hajto commented on test/adapters/websocex_test.exs:
```diff
@@ -0,0 +1,98 @@
+        assert {:ok, connection} = Adapter.WebSockex.connect(url, self(), [])
+      end
+
+      test "return error on invalid url", %{} do
```
```suggestion
      test "return error on invalid url" do
```

Hajto commented on lib/adapter/adapter.ex:
```diff
@@ -0,0 +1,83 @@
+  - `c:Janus.Transport.WS.Adapter.disconnect/1`
+
+  Every adapter should have its internal state as after `c:connect/3` callback
+  invocation `message_receiver` will not be passed again but must be remembered for incoming response messages from ws server.
```
```suggestion
  `message_receiver` value will be provided only during `c:connect/3` callback.
```

I would personally move this notice to `connect/3` callback definition.

Hajto commented on lib/adapter/adapter.ex:
```diff
@@ -0,0 +1,83 @@
+
+  Adapter is a module that takes part in communicating `Janus.Transport.WS` module
+  with lower level ws client (e.g. `:websockex`), sending and passing back messages, notifying about socket status change.
+  It has to implement all of given callbacks:
```
That is just plain to obvious. If someone use @behaviour `Janus.Transport.WS.Adapter` then compiler will spit exactly what you said.

Hajto replied:
You might put information here to put the behaviour. Or provide link to the WebSockex adapter you made here.

Hajto commented on README.md:
```diff
@@ -0,0 +1,21 @@
+# ElixirJanusTransportWs
```
You might as well put here information that this package is a transport layer for Elixir janus (put link here).

Hajto replied:
We should put a disclaimer here that this package is not released and it is experimental.

Hajto commented on README.md:
```diff
@@ -0,0 +1,21 @@
+Documentation can be generated with [ExDoc](https://github.com/elixir-lang/ex_doc)
+and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
+be found at [https://hexdocs.pm/elixir_janus_transport_ws](https://hexdocs.pm/elixir_janus_transport_ws).
+
```
We should at least provide license here at some point, not necessarily this pr.

Hajto commented on lib/adapter/adapter.ex:
```diff
@@ -0,0 +1,83 @@
+defmodule Janus.Transport.WS.Adapter do
+  @moduledoc """
+  The adapter specification.
```
This line does not seem to be necessary.

Hajto commented on lib/adapter/adapter.ex:
```diff
@@ -0,0 +1,83 @@
+  The adapter specification.
+
+  Adapter is a module that takes part in communicating `Janus.Transport.WS` module
+  with lower level ws client (e.g. `:websockex`), sending and passing back messages, notifying about socket status change.
```
```suggestion
  This module takes part in communicating `Janus.Transport.WS` module with lower-level WebSocket client (e.g. `:websockex`).
  
   It's responsible for sending and passing back messages, and notifying about socket status change.
```

Hajto commented on lib/adapter/adapter.ex:
````diff
@@ -0,0 +1,83 @@
+  Callback should notify message receiver about its status change with
+  ```elixir
+  notify_status(receiver, {:disconnected, "any arbitrary data"})
+  ```
````
```suggestion
  The callback should notify message receiver about its status change with `{:disconnected, "any arbitrary data"}`
```

Hajto commented on lib/adapter/adapter.ex:
````diff
@@ -0,0 +1,83 @@
+  notify_status(receiver, {:disconnected, "any arbitrary data"})
+  ```
+
+  ## Arguments
````
I think that these types of argument definitions are superfluous. If you have an important not obvious information, e.g. payload that payload must be already binary encoded binary (although it could be inferred from type) you should put it as a paragraph in the `@doc` and not spell out each argument in the docs explicitly.

Hajto commented on lib/adapter/adapter.ex:
```diff
@@ -0,0 +1,83 @@
+defmodule Janus.Transport.WS.Adapter do
+  @moduledoc """
+  The adapter specification.
+
```
You can also put an example usage here similar to what WebSockex author did (https://github.com/Azolo/websockex/blob/master/lib/websockex.ex#L11)

Hajto commented on lib/adapter/websockex.ex:
```diff
@@ -0,0 +1,84 @@
+              {:error, :connection_timeout}
+          end
+
+        error ->
```
I would put `{:error, _} = error` (or whatever correct shape).

Hajto commented on lib/adapter/websockex.ex:
```diff
@@ -0,0 +1,84 @@
+
+    @impl true
+    def disconnect(client) do
+      Kernel.send(client, :disconnect)
```
You can use only `send/2`. You don't have to specify `Kernel`. The whole module is automatically imported.

Qizot replied:
Yes, and this is the problem here. `Adapter` module declares `send` callback and it conflicts with kernel implementation.

Hajto commented on lib/transport/ws.ex:
```diff
@@ -0,0 +1,69 @@
+defmodule Janus.Transport.WS do
+  @moduledoc """
+  Implements `Janus.Transport` behaviour for connecting with Janus gateway via websocket.
+
```
Provide example usage here.

Hajto commented on lib/transport/ws.ex:
```diff
@@ -0,0 +1,69 @@
+  `{url, adapter, opts}`, where:
+  * `url` - is a valid URL for connecting with gateway
+  * `adapter` - is a module implementing behavior of `Janus.Transport.WS.Adapter`
+  * `opts` - arbitrary options specific to adapters
```
```suggestion
  * `url` - a valid URL for connecting with gateway
  * `adapter` - a module implementing behavior of `Janus.Transport.WS.Adapter`
  * `opts` - arbitrary options specific to adapters
```

Hajto commented on lib/transport/ws.ex:
```diff
@@ -0,0 +1,69 @@
+    with {:ok, payload_parsed} <- Jason.decode(payload) do
+      {:ok, payload_parsed, s}
+    else
+      {:error, reason} ->
```
If this failed to decode incoming stuff it means that either Janus messed up or the adapter failed to receive a frame. I think both of those cases deserve a log. (at warn level probably) Otherwise, you could do the `!` version and let it crash. We can't recover from reading the frame wrong I believe.

Qizot replied:
I decided to log warning.

Hajto commented on test/adapters/websocex_test.exs:
```diff
@@ -0,0 +1,98 @@
+if Code.ensure_loaded?(WebSockex) do
+  defmodule WebSocexTest do
```
```suggestion
  defmodule Janus.Transport.WS.Adapter.WebSockexTest do
```

Hajto commented on test/adapters/websocex_test.exs:
```diff
@@ -0,0 +1,98 @@
+      TestWebSocket.Server.start()
+    end
+
+    describe "websockex adapter should" do
```
Add setup here starting connection.
I might also add here a comment that this suite tests adapter layer only against websocket cowboy server.

Hajto commented on test/support/fake_ws_adapter.ex:
```diff
@@ -0,0 +1,81 @@
+  def start_link(url, message_receiver, opts) do
+    GenServer.start_link(
+      __MODULE__,
+      [url: url, message_receiver: message_receiver, opts: opts],
```
move opts to a variable so this invocation fits into 2 lines

Hajto commented on test/support/websocket_server/socket_handler.ex:
```diff
@@ -0,0 +1,33 @@
+      _ref -> nil
+    end
+
+    {:cowboy_websocket, request, state}
```
```suggestion
    {:cowboy_websocket, request, state}
```
`TestWebSocket.ClientConnection.store/1` seems to be an agent and does not utilize ETS there is no point in initializing ets here.

Hajto replied:
I would expect `TestWebSocket.ClientConnection.start_link` to be called here instead. I may be wrong though. It is started manually in the setup clause so after all, you might just want to delete that ets init thing.

Qizot replied:
Yup, I just forgot to delete this ets fragment.

Hajto commented on test/transport/transport_test.exs:
```diff
@@ -0,0 +1,61 @@
+
+  describe "send should" do
+    test "return ok on successfuly sent message" do
+      {:ok, state} = WS.connect({@fake_url, @adapter, []})
```
You have `WS.connect` in the setup_all clause, are you sure It is necessary to call it in every test?

Hajto commented on test/support/websocket_server/server.ex:
```diff
@@ -0,0 +1,22 @@
+        pid -> pid
+      end
+
+    %{server: cowboy_server, url: get_url(port)}
```
are you sure this shouldn't follow `{:ok, value}` | `{:error, reason}` pattern? It doesn't make sense even in tests.
If you put following code in the test code:
```elixir
{:ok, props} = TestWebSocket.server
```
you will immediately and clearly know what messed up. If used current implementation you would a bunch of errors that `{:error, :you_messed_up}` does not fit clause in couple of places in a lot of places. Error following the pattern that said error does not fit `{:ok, something}` is much clearer to read for me.

Hajto commented on test/support/websocket_server/server.ex:
```diff
@@ -0,0 +1,22 @@
+defmodule TestWebSocket.Server do
+  def get_url(port) do
```
If this function is public, why would you return the URL as part of the other function (which does not give much sense context wise).

Hajto replied:
```suggestion
  @default_port 8081
  def get_url(port \\ @default_port) do
```

Hajto commented on test/support/websocket_server/server.ex:
```diff
@@ -0,0 +1,22 @@
+    "ws://localhost:#{port}/ws"
+  end
+
+  def start(port \\ 8081) do
```
```suggestion
  def start(port \\ @default_port) do
```

Hajto commented on test/support/websocket_server/router.ex:
```diff
@@ -0,0 +1,23 @@
+       ]}
+    ]
+
+    port = opts[:port] || 4000
```
If you have running let's say "default phoenix application that uses janus" in the background (for example in docker container that does run constantly in the background) this would fail in the default clause.

Qizot replied:
is port like `8081` better in this case?

Hajto commented on test/support/fake_ws_adapter.ex:
```diff
@@ -0,0 +1,81 @@
+      message_receiver: args[:message_receiver],
+      opts: args[:opts],
+      message_fail: args[:opts][:message_fail] || false
+    }
```
```suggestion
    state = %{
      url: args[:url],
      message_receiver: args[:message_receiver],
      opts: args[:opts]
    }
```

Hajto replied:
If you accept this change then in almost all tests you can use one server and start it in setup clause.

Hajto commented on test/support/fake_ws_adapter.ex:
```diff
@@ -0,0 +1,81 @@
+      {:reply, {:error, "send error"}, state}
+    else
+      {:reply, :ok, state}
+    end
```
```suggestion
  def handle_call({:send, payload}, _from, state) do
    if payload == "fail" do
      {:reply, {:error, "send error"}, state}
    else
      {:reply, :ok, state}
    end
```

Hajto commented on mix.exs:
```diff
@@ -0,0 +1,45 @@
+
+  defp deps do
+    [
+      {:bunch, "~> 1.3"},
```
If you end up removing that withl I think you can remove this.

Hajto commented on test/support/websocket_server/server.ex:
```diff
@@ -0,0 +1,22 @@
+
+  def start(port \\ 8081) do
+    cowboy_server =
+      Plug.Cowboy.http(
```
I would suggest making sure that this call either links that servers or make sure its dead by the end of the test suite wherever it is used.

Qizot replied:
Added `on_exit` inside `setup_all` function.

Hajto commented on test/adapters/websocex_test.exs:
```diff
@@ -0,0 +1,98 @@
+        msg =
+          receive do
+            {:disconnected, info} -> {:disconnected, info}
+          end
```
```suggestion
       assert_receive {:disconnected, _info}  = msg
```
You can try this in test suite to make sure it works.
```elixir
send(self, {:e, :b})
assert_receive {:e, _info} = message
IO.inspect(message)
```

Hajto commented on test/adapters/websocex_test.exs:
```diff
@@ -0,0 +1,98 @@
+            {:disconnected, info} -> {:disconnected, info}
+          end
+
+        {:stop, {:disconnected, _}, _} = WS.handle_info(msg, state)
```
This test does not contain assert, are you missing one? Also, this could be unit tested, without server running.

Hajto commented:
Also setup CI

Qizot committed:
Adjusted to review part one

Qizot committed:
Adjusted to review and setup CI

Qizot committed:
Fixed inconsistent tests

just use setup_all to create single websocket server for all tests
instead of creating new one on each test

Qizot committed:
Removed redundtant connect in test

Hajto commented on .circleci/config.yml:
```diff
@@ -0,0 +1,23 @@
+  version: 2
+  build:
+    jobs:
+      - test
```
Missing new line at the end of the file

Hajto commented on .circleci/config.yml:
```diff
@@ -0,0 +1,23 @@
+      - checkout
+      - run: mix deps.get
+      - run: mix format --check-formatted
+      - run: mix compile --force
```
warnings as errors

Hajto commented on lib/adapters/adapter.ex:
```diff
@@ -0,0 +1,100 @@
+    @impl true
+    def connect(url, receiver, _opts) do
+      {:ok, fake_socket} = Agent.start_link(fn -> receiver end)
+      {:ok, fake_socket}
```
I know its for clarity sake, but this double `{:ok, fake_socket}` looks hilarious.

Hajto commented on lib/adapters/adapter.ex:
```diff
@@ -0,0 +1,100 @@
+  end
+
+  @doc """
+  Helper funciton to notify given receiver with connection status change.
```
```suggestion
  Helper function to notify the receiver with connection status change.
```

Hajto commented on lib/adapters/adapter.ex:
```diff
@@ -0,0 +1,100 @@
+  end
+
+  @doc """
+  Helper function to forward message received via websocket to message reciever previously initialized during `c:connect/3`.
```
```suggestion
  Helper function to forward the message received via websocket to message receiver previously initialized during `c:connect/3`.
```

Hajto commented on lib/adapters/adapter.ex:
```diff
@@ -0,0 +1,100 @@
+  @doc """
+  Closes given socket connection on demand.
+
+  The calblack should notify message receiver about its status change with `{:disconnected, "any arbitrary data"}` message.
```
```suggestion
  The callback should notify the message receiver about its status change with `{:disconnected, "any arbitrary data"}` message.
```

Hajto commented on lib/adapters/adapter.ex:
```diff
@@ -0,0 +1,100 @@
+  @type status_receiver_t :: pid()
+
+  @doc """
+  Creates new websocket connection.
```
```suggestion
  Creates a new websocket connection.
```

Hajto commented on lib/adapters/websockex.ex:
```diff
@@ -0,0 +1,84 @@
+
+    @impl true
+    def send(client, payload) do
+      if Process.info(client) do
```
The topic for discussion. Should you be defensive here or catch an exception as the library author expected?

Qizot replied:
I made a mistake here. As far as I remember in tests I was trying to send message via client that has already exited which should not happen (because we should have been notified about the exit). The library returns either `:ok` or tuple with an error and it should be used instead.

Hajto commented on lib/adapters/websockex.ex:
```diff
@@ -0,0 +1,84 @@
+
+    @impl true
+    def disconnect(client) do
+      Kernel.send(client, :disconnect)
```
```suggestion
      send(client, :disconnect)
```
I am pretty sure I've risen this topic on this repo before.

Qizot replied:
I shouldn't have resolved that. Adapter has `send` callback which interferes with Kernel's function.

Hajto commented on lib/adapters/websockex.ex:
```diff
@@ -0,0 +1,84 @@
+      Kernel.send(client, :disconnect)
+    end
+
+    def start_link(url, state) do
```
Is this expected to be used in different modules? At first glance, this looks like a private function that you forgot `p`.

Hajto commented on lib/adapters/websockex.ex:
```diff
@@ -0,0 +1,84 @@
+        {:ok, ws} ->
+          # process have started but connection may still not be made
+          # therefore wait for response from handle_connect callback
+          receive do
```
maybe that logic should be part of start_link?

Hajto commented on lib/adapters/websockex.ex:
```diff
@@ -0,0 +1,84 @@
+            timeout ->
+              # ws might still try to connect, kill so it can stop
+              # websockex has no option to cancel connection (from what I've tried to find)
+              Process.exit(ws, :kill)
```
Are you sure sending kill is mandatory?

Qizot replied:
Right, it's not. Changed to `close` function.

Hajto commented on lib/transport/ws.ex:
```diff
@@ -0,0 +1,85 @@
+  * `opts` - arbitrary options specific to adapters
+
+
+  ## Example
```
Not 100% sure this indentation is correct...

Qizot replied:
Do you mean this 2 blank lines or `iex>` formatting below? It looks alright on generated docs.

Hajto commented on lib/transport/ws.ex:
```diff
@@ -0,0 +1,85 @@
+  def connect({url, adapter, opts}) do
+    with {:ok, connection} <- adapter.connect(url, self(), opts) do
+      {:ok,
+       state(
```
Are you sure we can't fit this 5 lines into 2?

Hajto commented on lib/transport/ws.ex:
```diff
@@ -0,0 +1,85 @@
+    else
+      {:error, reason} -> {:error, {:send, reason}, s}
+    end
+  end
```
```suggestion
  @impl true
  def send(
        payload,
        _timeout,
        state(connection: connection, adapter: adapter) = state
      ) do
    payload = Jason.encode!(payload)

    with :ok <- adapter.send(connection, payload) do
      {:ok, state}
    else
      {:error, reason} -> {:error, {:send, reason}, state}
    end
  end
```

Hajto commented on lib/transport/ws.ex:
```diff
@@ -0,0 +1,85 @@
+
+        {:stop, {:parse_failed, payload, reason}, s}
+    end
+  end
```
```suggestion
  def handle_info({:ws_message, payload}, state) do
    with {:ok, payload_parsed} <- Jason.decode(payload) do
      {:ok, payload_parsed, state}
    else
      {:error, reason} ->
        Logger.warn(
          "[ #{__MODULE__} ] failed to parse incomming message with reason: #{inspect(reason)}"
        )

        {:stop, {:parse_failed, payload, reason}, state}
    end
  end
```

Hajto commented on mix.exs:
```diff
@@ -0,0 +1,52 @@
+defmodule ElixirJanusTransportWs.MixProject do
```
I would consider fixing this name.

Hajto commented on mix.exs:
```diff
@@ -0,0 +1,52 @@
+      {:ex_doc, "~> 0.22", only: :dev, runtime: false},
+
+      # TESTING
+      {:mock, "~> 0.3.0", only: :test},
```
I couldn't find mock used anywhere. Make sure it is actually used or remove it.

Hajto commented on test/support/fake_ws_adapter.ex:
```diff
@@ -0,0 +1,79 @@
+    state = %{
+      url: args[:url],
+      message_receiver: args[:message_receiver],
+      opts: args[:opts]
```
Those opts are not used anymore, I know this is only test server, but it does not make sense to me to keep them in state at this particular moment.

Qizot committed:
Apply suggestions from code review

Co-authored-by: Haito <hajtosek@gmail.com>

Qizot committed:
Adjusted to review

Qizot committed:
Merge branch 'develop' of github.com:software-mansion-labs/elixir-janus-transport-ws into develop

Qizot committed:
Added licence and updated readme

bblaszkow06 commented on .circleci/config.yml:
```diff
@@ -0,0 +1,25 @@
+      - run: mix format --check-formatted
+      - run: mix compile --force --warnings-as-errors
+      - run: mix test
+      - run: mix docs 2>&1 | (! grep -q "warning:")
```
I've found a way to both print and search for warnings:
```suggestion
      - run: mix docs 2>&1 | tee >(! grep -q "warning:")
```

bblaszkow06 commented on lib/adapters/adapter.ex:
```diff
@@ -0,0 +1,99 @@
+defmodule Janus.Transport.WS.Adapter do
+  @moduledoc """
+  This module takes part in communicating `Janus.Transport.WS` module
+  with lower level WebSocket client (e.g. `:websockex`).
```
```suggestion
  This module specifies the behaviour for adapter modules communicating `Janus.Transport.WS` module
  with lower level WebSocket client (e.g. `:websockex`).
```

bblaszkow06 commented on lib/adapters/adapter.ex:
```diff
@@ -0,0 +1,99 @@
+  This module takes part in communicating `Janus.Transport.WS` module
+  with lower level WebSocket client (e.g. `:websockex`).
+
+  It is responsible for sending and passing back messages, notifying about socket status change.
```
```suggestion
  An adapter is responsible for sending and passing WebSocket frames and notifying about socket status change.
```

bblaszkow06 replied:
The change regarding `messages` applies to the whole PR - WebSockets exchange frames, and if possible, use "message" only to reference erlang messages

bblaszkow06 commented on lib/adapters/adapter.ex:
```diff
@@ -0,0 +1,99 @@
+
+  It is responsible for sending and passing back messages, notifying about socket status change.
+
+  Sending messages is supposed to be synchronous while receiving is asynchronous.
```
Why sending has to be synchronous? `:gun`, the first library that may receive adapter after websockex, provides only `ws_send` that uses `cast` underneath.

Qizot replied:
Otherwise WS transport will never return error on message send when called from `Connection` module. Therefore we will need to handle asynchronous error in some other place where it will not be so obvious where it came from and from which call. If those are not problems then it might as well be async.

Qizot replied:
Changed it to be async.

bblaszkow06 commented on lib/adapters/adapter.ex:
```diff
@@ -0,0 +1,99 @@
+      :ok = Agent.stop(fake_socket)
+      :ok
+    end
+  end
```
This example is not useful. Try to do something like Tesla: https://hexdocs.pm/tesla/Tesla.Adapter.html#module-example with pseudocode, but closer to real life.

bblaszkow06 commented on lib/adapters/adapter.ex:
```diff
@@ -0,0 +1,99 @@
+
+  @type websocket_t :: pid()
+  @type url_t :: String.t()
+  @type payload_t :: binary()
```
For performance reasons, IO lists are often used in such context. It should be allowed here as well. If `:websockex` does not support that, conversion might be needed  (`IO.iodata_to_binary`)

bblaszkow06 commented on lib/adapters/adapter.ex:
```diff
@@ -0,0 +1,99 @@
+  @callback send(websocket :: websocket_t(), payload :: payload_t()) :: :ok | {:error, any}
+
+  @doc """
+  Closes given socket connection on demand.
```
```suggestion
  Closes the WebSocket connection.
```

bblaszkow06 commented on lib/adapters/adapter.ex:
```diff
@@ -0,0 +1,99 @@
+  @doc """
+  Closes given socket connection on demand.
+
+  The callback should notify the message receiver about its status change with `{:disconnected, "any arbitrary data"}` message.
```
Be more specific about "any arbitrary data" - what is it for? Is it some kind of description or disconnection reason? If it can actually be anything, do not suggest it's a binary but rather put a variable name there

bblaszkow06 commented on lib/adapters/websockex.ex:
```diff
@@ -0,0 +1,90 @@
+    defp start_link(url, timeout, args) do
+      websockex_opts = [
+        extra_headers: [{"Sec-WebSocket-Protocol", "janus-protocol"}]
+      ]
```
This header will be common for all adapters. The Adapter API should allow passing headers from connection and the role of adapter would be to put it under proper key/list when calling library

bblaszkow06 commented on lib/transport/ws.ex:
```diff
@@ -0,0 +1,78 @@
+defmodule Janus.Transport.WS do
+  @moduledoc """
+  Implements `Janus.Transport` behaviour for connecting with Janus gateway via websocket.
```
```suggestion
  Implements `Janus.Transport` behaviour for connecting with Janus gateway via WebSocket.
```

bblaszkow06 commented on lib/adapters/adapter.ex:
```diff
@@ -0,0 +1,99 @@
+  end
+
+  @doc """
+  Helper function to notify the receiver with connection status change.
```
What is connection status? If Connection is supposed to handle statuses, the allowed list has to be clearly specified and documented.

bblaszkow06 commented on lib/transport/ws.ex:
```diff
@@ -0,0 +1,78 @@
+        _timeout,
+        state(connection: connection, adapter: adapter) = state
+      ) do
+    payload = Jason.encode!(payload)
```
As mentioned above, you can use `encode_to_iodata`. Excerpt from docs:

>This function should be preferred to encode/2, if the generated JSON will be handed over to one of the IO functions or sent over the socket. The Erlang runtime is able to leverage vectorised writes and avoid allocating a continuous buffer for the whole resulting string, lowering memory use and increasing performance.

bblaszkow06 commented on lib/adapters/websockex.ex:
```diff
@@ -0,0 +1,90 @@
+    use Janus.Transport.WS.Adapter
+    use WebSockex
+
+    @impl true
```
You're implementing 2 behaviours, so you should rather use `@impl Adapter` and `@impl WebSockex`

Qizot committed:
Apply suggestions from code review

Co-authored-by: Bartosz Błaszków <bartosz.blaszkow@swmansion.com>

Qizot committed:
Adjusted to review

bblaszkow06 commented on lib/adapters/websockex.ex:
```diff
@@ -31,17 +37,17 @@ if Code.ensure_loaded?(WebSockex) do
-      websockex_opts = [
-        extra_headers: [{"Sec-WebSocket-Protocol", "janus-protocol"}]
-      ]
+      %{extra_headers: extra_headers} = args
```
Such args are usually Keyword, not map. Anyway, you should probably use `Access` behaviour (`args[:extra_headers]`)

bblaszkow06 commented on lib/adapters/adapter.ex:
```diff
@@ -1,76 +1,89 @@
-      receiver = Agent.get(fake_socket, fn receiver -> receiver end)
+    def disconnect(websocket) do
+      receiver = get_receiver(websocket)
+      :ok = disconnect_websocket(websocket)
```
This example showed a few more things to change:
- `websocket` is actually a state of the adapter, could be something else than pid (e.g. a map with receiver pid and WS pid)
- as state, `websocket` should be the last argument in callbacks
- `handle_frame` is quite confusing here (only when you know the same module is passed as a callback module to library it starts to make sense)

I'd make this example more abstract but more readable I think (feel free to adjust the following suggestion)
```elixir
defmodule CustomAdapter do
  use Janus.Transport.WS.Adapter

  @impl true
  def connect(url, receiver, opts) do
    on_receive = fn msg -> forward_message(receiver, msg) end
    ws_pid = SomeLibrary.spawn_websocket_connection(url, on_receive, opts)
    %{websocket: ws_pid, receiver: receiver}
  end

  @impl true
  def send(payload, %{websocket: pid}) do
    SomeLib.send_frame(pid, payload)
   end

  @impl true
  def disconnect(state) do
    :ok = SomeLib.disconnect(state.pid)
    notify_status(state.receiver, {:disconnected, "Disconnected request"})
  end
end
```

bblaszkow06 commented on lib/adapters/websockex.ex:
```diff
@@ -31,17 +37,17 @@ if Code.ensure_loaded?(WebSockex) do
+      opts = [extra_headers: extra_headers]
 
-      case WebSockex.start_link(url, __MODULE__, args, websockex_opts) do
+      case WebSockex.start_link(url, __MODULE__, args, opts) do
```
As mentioned, extract the implementation of Websockex callback module (can be in the same file) to have clear separation between 2 processes ( now we have callbacks executed in different context in one module)

Qizot committed:
Refactored websockex adapter

bblaszkow06 commented on lib/adapters/adapter.ex:
```diff
@@ -78,7 +62,7 @@ defmodule Janus.Transport.WS.Adapter do
   Sends payload via given WebSocket.
   """
-  @callback send(websocket :: websocket_t(), payload :: payload_t()) :: :ok | {:error, any}
+  @callback send(payload :: payload_t(), websocket :: websocket_t()) :: :ok | {:error, any}
```
I don't see type spec changes

bblaszkow06 replied:
1. websocket ->  state
2. modify the type (`websocket_t` does not have to be pid)

bblaszkow06 commented on lib/adapters/websockex.ex:
```diff
@@ -4,93 +4,101 @@ if Code.ensure_loaded?(WebSockex) do
-      {:ok, state}
-    end
+      args = %{
+        on_receive: fn frame -> forward_frame(receiver, frame) end,
+        on_disconnect: fn reason -> notify_status(receiver, {:disconnected, reason}) end,
```
I think you overcomplicated that a bit - there's no need to make this inner module so generic - treat it as private for this module and put those calls inside it (you can pass just a receiver)

bblaszkow06 commented on lib/adapters/websockex.ex:
```diff
@@ -4,93 +4,101 @@ if Code.ensure_loaded?(WebSockex) do
-    def connect(url, message_receiver, opts) do
-      timeout = opts[:timeout] || 5000
-      extra_headers = opts[:extra_headers] || []
+    defmodule WebSocketConnection do
```
```suggestion
    defmodule WebSocketConnection do
      @moduledoc false
```

Qizot committed:
Fixed specs
